### DISCUSSION TITLE
Is `mol.ToBinary()` guaranteed to be the same across RDKit version?

### DISCUSSION THREAD

**Discussion by hadim** [rdkit/rdkit#4919] [Q&A] [answered]
I am trying to build a unique ID taking into account tautomerism (two tautomeric forms should have a different ID) with the following:

```python
import datamol as dm
import hashlib


def generate_unique_id(smiles: str):
    mol_bi = dm.to_mol(smiles).ToBinary()
    mol_id = hashlib.md5(mol_bi).hexdigest()
    return mol_id
```

For this to work across RDKit version, I need to know whether `.ToBinary()` is guaranteed to be the same whatever version of RDKit used. My guess is that the binary representation might change, but I would like to double-check whether this is correct.

Note that for the above function to work, the atom ordering also must be the same (by using a canonical SMILES for example).

**Comment by greglandrum** [comment, marked as answer]
You cannot safely use the results of `.ToBinary()` to hash across RDKit versions.
I would suggest using the canonical SMILES and, if you have concerns about the SMILES changing from one version to the next, regenerate the SMILES each time you update the RDKit.

**Reply by hadim** [reply]
That was my guess. Thanks @greglandrum !

**Reply by hadim** [reply]
Did the SMILES canonical algorithm in RDKit every changed?

**Reply by greglandrum** [reply]
Yes. And bug fixes can leads to changes in the output from version to version. As I said, if you really need to be sure about the SMILES remaining the same you either need to stick to the same version or regenerate the SMILES whenever you update. This is just good practice: if the stability of those keys is important to you then you should regenerate them (or at least test that they aren’t change) anytime you change whichever software you are using to generate them.

**Reply by hadim** [reply]
Thanks for the clarification. Indeed that would be best to have a unique ID that can "survive" across RDKit versions, but I understand why it's not easily possible.

**Comment by hadim** [comment]
For the record, I found that it's possible to specify an InChi option during the generation process to fix the hydrogens `/FixedH`. It makes the InChi and InChiKey **non-standard** but can differentiate between tautomers: `Chem.inchi.MolToInchiKey(mol, options="/FixedH")`. See https://www.inchi-trust.org/technical-faq-2/#15.24 for more details and also more **non-standard** options.

xref with datamol where I plan to add wrappers for this: https://github.com/datamol-org/datamol/issues/74

_If I insist saying **non-standard**, it's because mixing standard and non-standard InChi is likely a bad idea so to be used with cautious!_

**Comment by greglandrum** [comment]
It's something of an edge case, but is worth being aware of: this does not stop InChI from rearranging charges.
```
In [8]: Chem.MolToInchi(Chem.MolFromSmiles('CN1C2=NC=CC=C2C(C)(C)\C1=C\C=C\C=C\C=C\C1=[N+](C)C2=CC=CC=C2C1(C)C'),options='/FixedH')
Out[8]: 'InChI=1/C28H32N3/c1-27(2)21-15-12-13-17-23(21)30(5)24(27)18-10-8-7-9-11-19-25-28(3,4)22-16-14-20-29-26(22)31(25)6/h7-20H,1-6H3/q+1'

In [9]: Chem.MolToInchi(Chem.MolFromSmiles('CN1C2=CC=CC=C2C(C)(C)\C1=C\C=C\C=C/C=C/C1=[N+](C)C2=NC=CC=C2C1(C)C'),options='/FixedH')
Out[9]: 'InChI=1/C28H32N3/c1-27(2)21-15-12-13-17-23(21)30(5)24(27)18-10-8-7-9-11-19-25-28(3,4)22-16-14-20-29-26(22)31(25)6/h7-20H,1-6H3/q+1'

In [10]: _8==_9
Out[10]: True
```

**Reply by hadim** [reply]
Interesting edge-case. Can you confirm this is a limitation of InChi itself and not RDKit?

**Reply by greglandrum** [reply]
I don’t see any reason why the RDKit would be responsible for this. Just try it with the InChI code and you can answer the question yourself.
It’s important to remember that the entire point of InChI is that it standardize the structure so that things which are duplicates within its chemistry model get marked as duplicates. It is not intended to be an interchange format
